atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/lock/LockCorrectnessChecker.java

```diff
@@ -0,0 +1,188 @@
+ *
+ * A successful unlock can be treated the same as a refresh, except that it has implications for further refreshes
+ * and unlocks (that we will check by other checkers).
+ *
+ * A successful lock means the lock was held <i>at some point</i> in the (closed) interval
+ * [a, b], where
+ * a is the InvokeEvent.time() of the lock request, and
+ * b is its OkEvent.time()
+ *
+ * We mantain a set L of all intervals where we know there was a lock, and a separate list U of all intervals where
+ * there was a lock at some point. In the latter list, we purposefully do not merge intervals, to ensure no information
+ * is lost.
+ *
+ * To verify locks were correctly granted, it is necessary and sufficient to verify that none of the uncertain
+ * intervals in U are fully covered by the intervals in the set of lock intervals L.
+ */
+public class LockCorrectnessChecker implements Checker {
+
+    private static final Logger log = LoggerFactory.getLogger(LockCorrectnessChecker.class);
+
+    @Override
+    public CheckerResult check(List<Event> events) {
+        Visitor visitor = new Visitor();
+        events.forEach(event -> event.accept(visitor));
+        visitor.verifyLockCorrectness();
+        return ImmutableCheckerResult.builder()
+                .valid(visitor.valid())
+                .errors(visitor.errors())
+                .build();
+    }
+
+    private static class Visitor implements EventVisitor {
+        private final Map<Pair<Integer, String>, InvokeEvent> pendingForProcessAndLock = new HashMap<>();
+        private final Map<Pair<Integer, String>, OkEvent> lastHeldLock = new HashMap<>();
+
+        private final Map<String, TreeRangeSet<Long>> locksHeld = new HashMap<>();
+        private final Map<String, List<Pair<InvokeEvent, OkEvent>>> locksAtSomePoint = new HashMap<>();
+
+        private final ArrayList<String> allLockNames = new ArrayList<>();
+
+        private final List<Event> errors = new ArrayList<>();
+        private final Map<Integer, String> resourceName = new HashMap<>();
+
+        @Override
+        public void visit(InfoEvent event) {
+        }
+
+        @Override
+        public void visit(InvokeEvent event) {
+            Integer process = event.process();
+            String lockName = event.value();
+            resourceName.put(process, lockName);
+
+            pendingForProcessAndLock.put(new Pair(process, lockName), event);
+            if (!allLockNames.contains(lockName)) {
+                allLockNames.add(lockName);
+                locksHeld.put(lockName, TreeRangeSet.create());
+                locksAtSomePoint.put(lockName, new ArrayList<>());
+            }
+        }
+
+        @Override
+        public void visit(OkEvent event) {
+            Integer process = event.process();
+            String lockName = resourceName.get(process);
+            Pair processLock = new Pair(process, lockName);
+            InvokeEvent invokeEvent = pendingForProcessAndLock.get(processLock);
```

Comment:
I think this assumes that an ok event must be preceded by the appropriate invoke event. Probably reasonable, though it would be nice to throw something a bit more helpful